### PR TITLE
[loader] Add loader contexts

### PR DESCRIPTION
**Pull request by lambdageek**
There are 4 different ways that an assembly can be loaded in .NET:
1. default context - `Assembly.Load(string)` does this.
2. reflection only context - `Assembly.ReflectionOnlyLoad` and `Assembly.ReflectionOnlyLoadFrom` do this.
2. load from context - `Assembly.LoadFrom(string)` does this. The interesting bit is that the assembly at the specified path is probed for its assembly name, and then binding redirects are applied to figure out if we need to load a different version of the assembly (from elsewhere in the search path, presumably).
3. "no context" (.NET Core's [AssemblyLoadContext](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md) design doc calls it "custom context", and internally they implement it as `IndividualAssemblyLoadContext`) - in Mono I'm going to call it the "individual context".  This is what `Assembly.LoadFile(string)` and `Assembly.Load(byte[])` do: you get exactly the thing you asked for even if something with the same assembly name is already loaded and you get to have problems with type resolution.

So this PR adds:
1. `MonoAssemblyLoadContextKind` enum - default, refonly, loadfrom and individual.
2. Store the `MonoAssemblyLoadContext` in every `MonoAssembly` instead of a boolean `refonly` flag.
3. When loading references from assemblies, consider the load context of the requesting assembly.
4. Do the wacky `LoadFrom` probe and rebind dance.
5. Don't check for duplicated for the individual assembly context.

Not in this PR:

The managed assembly load context stuff in .NET Core.  If we eventually want to add it, it will likely build on this work, but it's not here now.

Fixes: #8149, #8152, [bugzilla #11199](https://bugzilla.xamarin.com/show_bug.cgi?id=11199), [bugzilla #33728](https://bugzilla.xamarin.com/show_bug.cgi?id=33728)

TODO:
- [x] Add tests.  (I tried the repros from all the bugs above, but didn't turn them into tests yet).
- [x] Check that `ReflectionOnlyLoadFrom` is correct and binding redirects do not apply to it.
- [x] ~Find out what happens in this situation: directory `d` contains `Asm1.dll` version 1 and `Asm2.dll` version 1. App has a binding redirect for `Asm1` all versions to version 2.  normal app probing path has `Asm1` version 2 and `Asm2` version 2. `LoadFrom(d/Asm1.dll)` and then cause the `Asm2` reference to be loaded - do we get `Asm2` version 1 or version 2.  (ie did we retain `d` as a probing path even though the redirect made us look for `Asm1` somewhere else)~ it's fine... there's nothing that remembers that we started with a `LoadFrom` in `d/`. 
- [x] FIXME: resolving references from individual context assemblies should only fire the `AssemblyResolve` event and not look in the default context.
- [x] FIXME: SRE assemblies should be in the individual context.
- [x] FIXME: "assemblies that are loaded from byte arrays are loaded without context unless their identity (after policy is applied) establishes that they are in the global assembly cache." - I'm not applying policy or looking in the GAC...
- [x] Figure out the FIXMEs by trying .NET Framework
  - [x] ~If `LoadFrom` is subject to a binding redirect, where should `LoadFrom` look for the replacement assembly. Is the given filename's basename relevant at all? Is the new assembly loaded in the default context?~ No
  - [x] Should any sort of deduplication happen for `LoadFile` and `Load(byte[])` - should you really get brand new `MonoAssembly*` pointers every time (even if for `LoadFile` the `MonoImage` is shared).
  - [x] Am I firing all the right managed events? As best I can tell, yes.

### CODE REVIEW

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Hybrid Suspend

**Comment by akoeplinger** [comment]
> The managed assembly load context stuff in .NET Core. If we eventually want to add it, it will likely build on this work, but it's not here now.

FWIW I've seen e.g. Roslyn use reflection to probe for this type to check whether they're on .NET Core so we might need to coordinate this when we add it.

**Comment by lambdageek** [comment]
@kumpera It's definitely the case here that the code isn't making a distinction between `ReflectionOnlyLoad`and `ReflectionOnlyLoadFrom` - I'll rework the PR to add that.

@akoeplinger I'll make sure to coordinate if I start making changes to the managed interfaces.

**Comment by kumpera** [comment]
@lambdageek on refonly, we should validate my claim first. I'm not 100% sure on it.

**Comment by lambdageek** [comment]
Interesting. `ReflectionOnlyLoadFrom` in .NET Framework does _not_ have binding redirects applied. It  loads straight from the given path.  So what I have now is correct.

**Comment by lambdageek** [comment]
Changing `LoadFile` to use `MONO_ASMCTX_INDIVIDUAL` broke xUnit for some reason (changing it to `MONO_ASMCTX_DEFAULT` or even `LOADFROM` makes it work).  I can see that they use `LoadFile` on .NET Framework sometimes.  Investigating...

**Comment by lambdageek** [comment]
The reason xUnit broke is because apparently even for a `LoadFile`/"no context"/indivudal assembly context, the .NET Framework will still cause GAC assemblies (or at least framework Facade assemblies like `System.Runtime`) to be loaded (and won't fire an `AssemblyLoad` event to announce it, for some reason)).

That's... going to require a slight hack.  Mono doesn't really have a way of saying "Load from the GAC (and from MONO_PATH) but not from the current domain's paths").  So I'll have to add that.

**Comment by luhenry** [comment]
Looks like `Linux x64 FullAOT` is missing a AOTted assembly : https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-fullaot/10068/testReport/junit/MonoTests/runtime/assembly_loadfile_exe/

Do you think the `Linux AArch64 Interpreter` failure is related? https://jenkins.mono-project.com/job/test-mono-pull-request-interpreter-aarch64/4408/parsed_console/log_content.html#WARNING1. I'll retry it to check if it's a flake.

**Comment by luhenry** [comment]
@monojenkins build Windows x64

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 Interpreter

File: mono/mini/aot-runtime.c [hunk 1]
```diff
@@ -2326,7 +2347,10 @@ if (container_assm_name && !container_amodule) {
 		for (i = 0; i < amodule->image_table_len; ++i) {
 			ERROR_DECL (error);
 			load_image (amodule, i, error);
-			mono_error_cleanup (error); /* FIXME don't swallow the error */
+			if (!is_ok (error)) {
+				mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_AOT, "AOT: image '%s' could not load image %d of %d due to: %s", found_aot_name, i, amodule->image_table_len, mono_error_get_message (error));
+				mono_error_cleanup (error); /* FIXME don't swallow the error */
```
Comment (vargaz) [inline mono/mini/aot-runtime.c]: We already do a mono_trace from load_image ().

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux x64

**Comment by lambdageek** [comment]
@monojenkins build Linux i386

**Comment by marek-safar** [comment]
@mhutch no, we didn't test that with MD yet. We do that regularly during monthly integration work but if you have a specific concern we could do that earlier

**Comment by lambdageek** [comment]
@mhutch interesting.

I'll try VSfM on top of mono master today.  What's something specific I should try that would demonstrate the problem?  Open VS Mac and try to install some extensions?

**Comment by KirillOsenkov** [comment]
@lambdageek: @davkean and @marklio are my go to people for assembly loading, fusion, loader contexts, etc. in .NET Framework. 

Guys, who would be the best person to give this a sanity check? I know assembly loading is very tricky and this has potential to break stuff since I know Mono used to be much more lenient and I know people took dependency on the old behavior.

**Pull request merged**